javascript-review.js

```diff
@@ -4,117 +4,236 @@
+  }
+};
+
+coolObject.myFunction()
+
 // 2.
 
 // Write a constructor function (to create objects) that takes one argument. It should use that argument to set the value of the first attribute of the object. There should be a second attribute that stores the value ‘red’. 
 
 // Note that a constructor function is the old school syntax that was used in javascript before the sugar class syntax came along. 
 
+const objectConstruct = (attribute) => {
+  // object = {};
+  this.attribute = attribute;
+  this.secondAttribute = 'red';
+  return this;
+}
+
+myThing = objectConstruct('yes');
+console.log(myThing);
 // 3.
 
 // Can you access variables defined outside of functions within the scope of functions in JS? Show a simple example of this being used? (Define a constant in global scope and use that in a very simple function.)
 
+// Yes
+let bestGlobal = 'Hello';
+const scoping = () => {
+  console.log(bestGlobal + ' ' + 'World!');
+}
+scoping();
 // 4.
 
 // What two ways can be used to access the properties of objects in JS? Give examples of both (although assume the objects have been defined).
 
+object1 = {
+  thing1: 1,
+  thing2: 2,
+  thing3: 3
+}
+
+console.log(object1.thing1);
+console.log(object1['thing1']);
 // 5.
 
 // What is JSON? Be clear but concise.
 
+// JSON is a data format used commonly in web api's that emulates the 'structure' of javascript objects. A JSON file is always a simple text document that must be parsed by some parser to convert it to proper javascript. It is a data formatting convention that organises data in an efficient and javascripty manner.
+
 // 6.
 
 // What is a callback?
 
+// Callback is when you call one function as an argument to another function including anonymous functions
+const logStuff = (stuff) => {
+  console.log(stuff);
+};
+
+const logStuffAndHi = (stuff) => {
+  console.log(stuff);
+  console.log('hi');
+}
+
+const coolFunc = (stuff, loggingFunc) => {
+  loggingFunc(stuff);
+}
+
+coolFunc('always', logStuff)
+coolFunc('all day', logStuffAndHi)
 // 7.
 
 // Define a function with that takes two number arguments and a callback. The function will add the two numbers and pass them to the callback as an argument. Now call this function and in the callback simply console.log out the argument that has been passed through.
 
+const sumLog = (num1, num2, aFunc) => {
+  aFunc(num1 + num2);
+}
+sumLog(4, 5, logStuff);
+
 // 8.
 
 // Why do we use promises?
 
+// Promises are used for getting data from a request. Promises allow for javascript to use asynchronous functions so the page loads other javascript functionality that does not require data from a promise. Without promises pages would wait until all resources are loaded before any javascript functionality occurs.
+
 // 9.
 
 // What are the two methods at our disposal if we have hold of a promise? What are the names of the functions that are related to these functions (that are called when the promise is fulfilled or fails).
 
+// .then and .catch
```

Comment:
when we're in a promise we have `resolve` and `reject`

correct with `.then` and `.catch`